### ISSUE TITLE
Compat interface to support different OSes

### ISSUE THREAD

**Issue by ThomasAdam** [jaagr/polybar#1529]
## Is your feature request related to a problem? Please describe.

No.

## Why does polybar need this feature?

In order to use `polybar` on different OSes, there's currently two possibilities:

1.  If the operating system has some kind of Linux-emulation mode (such as `FreeBSD`) then the `polybar` code needs only a few changes -- however, this only helps a very limited subset of modules in `polybar`.

2.  One has to write separate modules for that OS which is currently problematic as that's replacing the current linux-only module with something that's only specific to that OS, which is not ideal either.

What we need to do is come up with an established interface that allows for the following:

* Platform-agnostic means to define data gathering for platform-specific modules in `polybar` such as:
  * cpu
  * memory
  * temperature
  * networking/wireless

In the same way, this will also help break the tight coupling between the current modules and how they're both responsible for data gathering and data rendering.

## Describe the solution you'd like

It seems as though the current abstraction uses a mixture of templating and factory base classes for the different modules.

We could have the following layout:

```
src/
	common.hpp
	compat/
		linux/
			cpu.cpp
			mem.cpp
		freebsd/
			cpu.cpp
			mem.cpp
		macos/
			cpu.cpp
			mem.cpp
		...
```

Then get `CMake` to include the relevant file(s) depending on the OS detected.  `common.hpp` would be a header file to aggregate the OS-specific code, and hence, `common.hpp` would be the header which gets included in the main part of polybar so everything links though.

In terms of structure, not much would change now, but we could explicitly make the modules a complete factory class, hence:

```
class ModuleDataGatherer {
	public:
		virtual int getData(...) const = 0;
		virtual ~ModuleDataGatherer() {};
};

class ModuleCPUGatherer : public ModuleDataGatherer {
	public:
		virtual int getData(...) const { ... };
		virtual ~ModuleCPUGatherer() {};
};

class ModuleMemoryGatherer : public ModuleDataGatherer {
	public:
		virtual int getData(...) const { ... };
		virtual ~ModuleMemoryGatherer() {};
};
```

Each OS-specific module would then implement the above, and we would have a wrapper class to act as the module factory class, something like:

```
class ModuleFactory {
	public:
		enum ModuleType {
			ModuleCPUGatherer,
			ModuleMemoryGatherer,
		};

		...
};
```

Something that would handle each module -- I know we have this already, but if we go down this path, then we're explicitly separating out data from rendering, so we would need to have something similar for rendering as well -- we could reuse most of what we have already -- it's just a nice tidy-up.  We should also ensure that options per module are separated out so that we isolate all the common options together, and only let module-specific options be a part of the encapsulation.

I've only had a very quick poke around the overall code in `polybar` so I hope I've not overlooked anything obvious, but this approach -- or something like it -- would certainly help with OS portability.

**Comment by patrick96** [comment]
The outline looks good. I have a few thoughts about the structure and implementation details:

For the layout I would have suggested going the adapter way, as we did with alsa, network, and pulse:
```
include/
    adapters/
        ...
        cpu.hpp
        mem.hpp
        ...
src/
    adapters/
        ...
        cpu.cpp
        cpu_freebsd.cpp
        mem.cpp
        mem_freebsd.cpp
        ...
```
The memory module would `#include "adapters/mem.hpp"` and during compilation the correct `.o` file will be linked. I also don't want to have coupling between the modules, that's why I didn't use a `common.hpp` file. This might look a little messy if we add this for many OSes but the goal is to have each module be its own shared library and all its files are in the same folder, as discussed in #1409.

I don't see the point in having `ModuleDataGatherer`, this common supertype would never be used anywhere because each module would need the specific type, since the interface for these gatherers will often be more complex than `getData()`, for example the memory module would need at least this interface:
```cpp
    virtual uint64_t get_total() = 0;
    virtual uint64_t get_avail() = 0;
    virtual uint64_t get_swap_total() = 0;
    virtual uint64_t get_swap_avail() = 0;
```

I don't understand what the purpose of `ModuleFactory` should be. Would it be a complete replacement for the current module factory?

You said you would be working on this, correct?
Could you start implementing this separation for a single module only (whichever you think would be easiest) so that we can see what is really involved.
If this goes as friction free as it looks, we should definitely get this merged.

**Comment by patrick96** [comment]
Don't worry about it. You would have seen enough of the architecture once you started implementing ;)

**Comment by NBonaparte** [comment]
On a similar note, I'd like to abstract the rendering away from XCB/X11 to implement Wayland support (#414), but everything is so tightly coupled to XCB that a lot of things have to be rewritten.

**Comment by ThomasAdam** [comment]
@NBonaparte, sure. But that's a separate consideration, with a very different goal, and not one this change would help with.

**Comment by patrick96** [comment]
@NBonaparte Yes, an abstraction layer between our renderer and wayland/xorg would be nice. But this proposal here would be totally independent from our renderer and would only affect how modules gather their data. (The only place where they would coincide is in the `xbacklight` module)

**Comment by ercdude** [comment]
How is this going? Would love to contribute to it, if possible, and add support for freebsd. Is there any WIP?

**Comment by patrick96** [comment]
@ercdude AFAIK there isn't being any work done on this right now (maybe in #1937). I think we'll tackle this in two stages. First we try to separate the data gathering and processing logic from the module formatting logic (backend - frontend) and after that it should become easier to implement different backends for different distros.

If you want you could try your hand at try to separate out the backend logic into an adapter in one of the more straightforward modules (e.g. cpu) just as a proof of concept and we can discuss further designs from there.
